internal/validators/os/os_valdiator_test.go

```diff
@@ -0,0 +1,114 @@
+package os
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/canonical/gomaasclient/entity"
+	"github.com/go-logr/logr"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/validator-labs/validator-plugin-maas/api/v1alpha1"
+)
+
+type DummyMaaSAPIClient struct {
+	images []entity.BootResource
+}
+
+func (d *DummyMaaSAPIClient) ListOSImages() ([]entity.BootResource, error) {
+	return d.images, nil
+}
+
+func TestFindingBootResources(t *testing.T) {
+
+	type TestCase struct {
+		Name        string
+		ruleService *ImageRulesService
+		imageRules  []v1alpha1.OSImageRule
+		errors      []error
+		details     []string
+	}
+
+	testCases := []TestCase{
+		{
+			Name: "image is found in MaaS",
+			ruleService: NewImageRulesService(
+				logr.Logger{},
+				&DummyMaaSAPIClient{
+					images: []entity.BootResource{
+						{
+							Name:         "Ubuntu",
+							Architecture: "amd64/ga-20.04",
+						},
+					},
+				},
+			),
+			imageRules: []v1alpha1.OSImageRule{
+				{Name: "Image found", OSImages: []v1alpha1.OSImage{
+					{OSName: "Ubuntu", Architecture: "amd64/ga-20.04"},
+				}},
+			},
+			errors:  make([]error, 0),
+			details: make([]string, 0),
+		},
+		{
+			Name: "image is not found in MaaS",
+			ruleService: NewImageRulesService(
+				logr.Logger{},
+				&DummyMaaSAPIClient{
+					images: make([]entity.BootResource, 0),
+				}),
+			imageRules: []v1alpha1.OSImageRule{
+				{Name: "Image not found", OSImages: []v1alpha1.OSImage{
+					{OSName: "Ubuntu", Architecture: "amd64/ga-20.04"},
+				}},
+			},
+			errors:  []error{errors.New("failed to validate rule")},
+			details: []string{"OS image Ubuntu with arch amd64/ga-20.04 was not found"},
+		},
+		{
+			Name: "a few images are not found in MaaS",
+			ruleService: NewImageRulesService(
+				logr.Logger{},
+				&DummyMaaSAPIClient{
+					images: make([]entity.BootResource, 0),
+				}),
+			imageRules: []v1alpha1.OSImageRule{
+				{Name: "Image not found", OSImages: []v1alpha1.OSImage{
+					{OSName: "Ubuntu", Architecture: "amd64/ga-20.04"},
+				}},
+				{Name: "Image not found", OSImages: []v1alpha1.OSImage{
+					{OSName: "Ubuntu", Architecture: "amd64/ga-22.04"},
+				}},
+			},
+			errors: []error{errors.New("failed to validate rule"), errors.New("failed to validate rule")},
+			details: []string{
+				"OS image Ubuntu with arch amd64/ga-20.04 was not found",
+				"OS image Ubuntu with arch amd64/ga-22.04 was not found"},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.Name, func(t *testing.T) {
+			var allErrors []error
+			var allDetails []string
+
+			for _, rule := range tc.imageRules {
```

Comment:
Typically unit tests only invoke a single function. Please update this test to invoke `ReconcileMaasInstanceImageRules` instead. Then it'll invoke `ListOSImages` and `findBootResources` for you, so the test code will be cleaner and there'll be more coverage.

Also, the name of the test function should be `Test<name_of_function_being_tested>`, i.e., `TestReconcileMaasInstanceImageRules`.